File: sdk-java/hfcx-sdk-client/src/test/java/eg/gov/healthflow/hfcx/sdk/client/recipient/RecipientHandlerTest.java
package eg.gov.healthflow.hfcx.sdk.client.recipient;

import eg.gov.healthflow.hfcx.sdk.client.OutboundEncryptor;
import eg.gov.healthflow.hfcx.sdk.client.protocol.ProtocolHeaders;
import eg.gov.healthflow.hfcx.sdk.client.registry.ParticipantCert;
import eg.gov.healthflow.hfcx.sdk.client.testsupport.TestCerts;
import eg.gov.healthflow.hfcx.sdk.core.exception.AuthenticationException;
import eg.gov.healthflow.hfcx.sdk.core.exception.BusinessException;
import eg.gov.healthflow.hfcx.sdk.core.exception.ProtocolException;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.time.ZoneOffset;
import java.util.Map;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class RecipientHandlerTest {

    private static TestCerts.GeneratedCert recipient;
    private static OutboundEncryptor outboundEncryptor;

    @BeforeAll
    static void setUp() throws Exception {
        recipient = TestCerts.generate("payerco@hcx-egypt", Duration.ofDays(30));
        outboundEncryptor = new OutboundEncryptor(code -> new ParticipantCert(
                code, recipient.publicKey(), Instant.now().plusSeconds(3600)));
    }

    private RecipientHandler.Builder validBuilder() {
        return RecipientHandler.builder()
                .keyProvider(() -> recipient.privateKey())
                .localParticipantCode("payerco@hcx-egypt")
                .bearerTokenValidator(authHeader -> {
                    if (authHeader == null || !authHeader.startsWith("Bearer ")) {
                        throw new AuthenticationException("missing or malformed bearer");
                    }
                });
    }

    private static String envelope(String payload) {
        String jwe = outboundEncryptor.encrypt(payload, "payerco@hcx-egypt");
        return "{\"payload\":\"" + jwe + "\"}";
    }

    private static String validBundle() {
        return "{\"resourceType\":\"Bundle\",\"type\":\"collection\",\"entry\":["
                + "{\"resource\":{\"resourceType\":\"Patient\","
                + "\"identifier\":[{\"system\":\"" + FhirValidator.NATIONAL_ID_SYSTEM
                + "\",\"value\":\"29504150112355\"}],"
                + "\"address\":[{\"country\":\"EG\"}]}}]}";
    }

    private static Map<String, String> validHeaders(String correlationId, Instant now) {
        return Map.of(
                ProtocolHeaders.SENDER_CODE, "myhospital@hcx-egypt",
                ProtocolHeaders.RECIPIENT_CODE, "payerco@hcx-egypt",
                ProtocolHeaders.CORRELATION_ID, correlationId,
                ProtocolHeaders.TIMESTAMP, now.toString(),
                ProtocolHeaders.API_CALL_ID, UUID.randomUUID().toString());
    }

    @Test
    void endToEndRoundTripFromOutboundEncryptorYieldsTheOriginalPayload() {
        RecipientHandler handler = validBuilder().build();
        String correlationId = UUID.randomUUID().toString();
        Instant now = Instant.now();
        String bundle = validBundle();

        RecipientResult result = handler.handle(
                "Bearer test-token",
                validHeaders(correlationId, now),
                envelope(bundle));

        assertEquals(bundle, result.decryptedPayload());
        assertEquals(correlationId, result.correlationId());
    }

    @Test
    void missingBearerHeaderRaisesAuthenticationException() {
        RecipientHandler handler = validBuilder().build();
        AuthenticationException ex = assertThrows(AuthenticationException.class, () ->
                handler.handle(null, validHeaders(UUID.randomUUID().toString(), Instant.now()),
                        envelope(validBundle())));
        assertEquals("ERR-T-002", ex.getCode());
    }

    @Test
    void disablingBearerLayerSkipsValidatorRequirement() {
        // Build with no validator AND disable the layer — should succeed.
        RecipientHandler handler = RecipientHandler.builder()
                .keyProvider(() -> recipient.privateKey())
                .localParticipantCode("payerco@hcx-egypt")
                .enable(Layer.BEARER, false)
                .build();
        RecipientResult result = handler.handle(null,
                validHeaders(UUID.randomUUID().toString(), Instant.now()),
                envelope(validBundle()));
        assertTrue(result.decryptedPayload().contains("Bundle"));
    }

    @Test
    void enablingBearerWithoutValidatorFailsAtConstruction() {
        IllegalStateException ex = assertThrows(IllegalStateException.class, () ->
                RecipientHandler.builder()
                        .keyProvider(() -> recipient.privateKey())
                        .localParticipantCode("payerco@hcx-egypt")
                        .build());
        assertTrue(ex.getMessage().contains("BearerTokenValidator"));
    }

    @Test
    void recipientCodeMismatchRaisesProtocolException() {
        RecipientHandler handler = validBuilder().build();
        Map<String, String> headers = Map.of(
                ProtocolHeaders.SENDER_CODE, "myhospital@hcx-egypt",
                ProtocolHeaders.RECIPIENT_CODE, "wrong-recipient@hcx-egypt",
                ProtocolHeaders.CORRELATION_ID, UUID.randomUUID().toString(),
                ProtocolHeaders.TIMESTAMP, Instant.now().toString(),
                ProtocolHeaders.API_CALL_ID, UUID.randomUUID().toString());
        ProtocolException ex = assertThrows(ProtocolException.class, () ->
                handler.handle("Bearer x", headers, envelope(validBundle())));
        assertEquals(HeaderValidator.CODE_RECIPIENT_MISMATCH, ex.getCode());
    }

    @Test
    void missingProtocolHeaderRaisesProtocolException() {
        RecipientHandler handler = validBuilder().build();
        Map<String, String> headers = Map.of(
                ProtocolHeaders.SENDER_CODE, "myhospital@hcx-egypt",
                // recipient_code missing
                ProtocolHeaders.CORRELATION_ID, UUID.randomUUID().toString(),
                ProtocolHeaders.TIMESTAMP, Instant.now().toString(),
                ProtocolHeaders.API_CALL_ID, UUID.randomUUID().toString());
        ProtocolException ex = assertThrows(ProtocolException.class, () ->
                handler.handle("Bearer x", headers, envelope(validBundle())));
        assertEquals(HeaderValidator.CODE_MISSING_HEADER, ex.getCode());
    }

    @Test
    void timestampOutsideToleranceRaisesProtocolException() {
        // Pin clock so the timestamp delta is deterministic.
        Clock fixedClock = Clock.fixed(Instant.parse("2026-05-07T12:00:00Z"), ZoneOffset.UTC);
        RecipientHandler handler = validBuilder().clock(fixedClock).build();
        Map<String, String> headers = validHeaders(UUID.randomUUID().toString(),
                Instant.parse("2026-05-07T11:50:00Z"));  // 10min in the past, > 5min tolerance
        ProtocolException ex = assertThrows(ProtocolException.class, () ->
                handler.handle("Bearer x", headers, envelope(validBundle())));
        assertEquals(HeaderValidator.CODE_TIMESTAMP_OUT_OF_RANGE, ex.getCode());
    }

    @Test
    void disablingHeadersLayerSkipsHeaderChecks() {
        RecipientHandler handler = validBuilder().enable(Layer.HEADERS, false).build();
        // Wrong recipient code: would normally raise; disabled layer means accepted.
        Map<String, String> headers = Map.of(
                ProtocolHeaders.SENDER_CODE, "myhospital@hcx-egypt",
                ProtocolHeaders.RECIPIENT_CODE, "wrong@hcx-egypt",
                ProtocolHeaders.CORRELATION_ID, UUID.randomUUID().toString(),
                ProtocolHeaders.TIMESTAMP, Instant.now().toString(),
                ProtocolHeaders.API_CALL_ID, UUID.randomUUID().toString());
        RecipientResult result = handler.handle("Bearer x", headers, envelope(validBundle()));
        assertTrue(result.decryptedPayload().contains("Patient"));
    }

    @Test
    void fhirLayerRejectsNonBundleResource() {
        RecipientHandler handler = validBuilder().build();
        String body = envelope("{\"resourceType\":\"Patient\"}");
        BusinessException ex = assertThrows(BusinessException.class, () ->
                handler.handle("Bearer x",
                        validHeaders(UUID.randomUUID().toString(), Instant.now()), body));
        assertEquals(FhirValidator.CODE_NOT_A_BUNDLE, ex.getCode());
    }

    @Test
    void fhirLayerRejectsPatientWithoutNationalIdIdentifier() {
        RecipientHandler handler = validBuilder().build();
        String bundle = "{\"resourceType\":\"Bundle\",\"type\":\"collection\",\"entry\":["
                + "{\"resource\":{\"resourceType\":\"Patient\","
                + "\"identifier\":[{\"system\":\"http://other.example.com/id\",\"value\":\"x\"}],"
                + "\"address\":[{\"country\":\"EG\"}]}}]}";
        BusinessException ex = assertThrows(BusinessException.class, () ->
                handler.handle("Bearer x",
                        validHeaders(UUID.randomUUID().toString(), Instant.now()),
                        envelope(bundle)));
        assertEquals(FhirValidator.CODE_PATIENT_MISSING_NATIONAL_ID, ex.getCode());
    }

    @Test
    void fhirLayerRejectsNonEgyptianAddress() {
        RecipientHandler handler = validBuilder().build();
        String bundle = "{\"resourceType\":\"Bundle\",\"type\":\"collection\",\"entry\":["
                + "{\"resource\":{\"resourceType\":\"Patient\","
                + "\"identifier\":[{\"system\":\"" + FhirValidator.NATIONAL_ID_SYSTEM
                + "\",\"value\":\"29504150112355\"}],"
                + "\"address\":[{\"country\":\"US\"}]}}]}";
        BusinessException ex = assertThrows(BusinessException.class, () ->
                handler.handle("Bearer x",
                        validHeaders(UUID.randomUUID().toString(), Instant.now()),
                        envelope(bundle)));
        assertEquals(FhirValidator.CODE_PATIENT_NON_EGYPTIAN, ex.getCode());
    }

    @Test
    void disablingFhirLayerSkipsStructuralValidation() {
        RecipientHandler handler = validBuilder().enable(Layer.FHIR, false).build();
        // Not even a Bundle — would normally be rejected.
        RecipientResult result = handler.handle("Bearer x",
                validHeaders(UUID.randomUUID().toString(), Instant.now()),
                envelope("{\"resourceType\":\"Patient\"}"));
        assertTrue(result.decryptedPayload().contains("Patient"));
    }

    @Test
    void egyptianLayerRejectsInvalidNationalIdValue() {
        RecipientHandler handler = validBuilder().build();
        String bundle = "{\"resourceType\":\"Bundle\",\"type\":\"collection\",\"entry\":["
                + "{\"resource\":{\"resourceType\":\"Patient\","
                + "\"identifier\":[{\"system\":\"" + FhirValidator.NATIONAL_ID_SYSTEM
                + "\",\"value\":\"00000000000000\"}],"
                + "\"address\":[{\"country\":\"EG\"}]}}]}";
        BusinessException ex = assertThrows(BusinessException.class, () ->
                handler.handle("Bearer x",
                        validHeaders(UUID.randomUUID().toString(), Instant.now()),
                        envelope(bundle)));
        assertEquals(EgyptianBundleValidator.CODE_BAD_NATIONAL_ID, ex.getCode());
    }

    @Test
    void egyptianLayerRejectsInvalidPhoneValue() {
        RecipientHandler handler = validBuilder().build();
        String bundle = "{\"resourceType\":\"Bundle\",\"type\":\"collection\",\"entry\":["
                + "{\"resource\":{\"resourceType\":\"Patient\","
                + "\"identifier\":[{\"system\":\"" + FhirValidator.NATIONAL_ID_SYSTEM
                + "\",\"value\":\"29504150112355\"}],"
                + "\"telecom\":[{\"system\":\"phone\",\"value\":\"01312345678\"}],"
                + "\"address\":[{\"country\":\"EG\"}]}}]}";
        BusinessException ex = assertThrows(BusinessException.class, () ->
                handler.handle("Bearer x",
                        validHeaders(UUID.randomUUID().toString(), Instant.now()),
                        envelope(bundle)));
        assertEquals(EgyptianBundleValidator.CODE_BAD_PHONE, ex.getCode());
    }

    @Test
    void disablingEgyptianLayerSkipsValueValidation() {
        RecipientHandler handler = validBuilder().enable(Layer.EGYPTIAN, false).build();
        String bundle = "{\"resourceType\":\"Bundle\",\"type\":\"collection\",\"entry\":["
                + "{\"resource\":{\"resourceType\":\"Patient\","
                + "\"identifier\":[{\"system\":\"" + FhirValidator.NATIONAL_ID_SYSTEM
                + "\",\"value\":\"00000000000000\"}],"
                + "\"address\":[{\"country\":\"EG\"}]}}]}";
        // Invalid National-ID — would normally raise; layer disabled means accepted.
        RecipientResult result = handler.handle("Bearer x",
                validHeaders(UUID.randomUUID().toString(), Instant.now()),
                envelope(bundle));
        assertTrue(result.decryptedPayload().contains("00000000000000"));
    }

    @Test
    void allLayersDisabledStillDecrypts() {
        RecipientHandler handler = RecipientHandler.builder()
                .keyProvider(() -> recipient.privateKey())
                .enable(Layer.BEARER, false)
                .enable(Layer.HEADERS, false)
                .enable(Layer.FHIR, false)
                .enable(Layer.EGYPTIAN, false)
                .build();
        RecipientResult result = handler.handle(null,
                Map.of(ProtocolHeaders.CORRELATION_ID, "x"),
                envelope("plain string payload"));
        assertEquals("plain string payload", result.decryptedPayload());
    }

    @Test
    void enabledLayersAccessorReflectsBuilderToggles() {
        RecipientHandler handler = validBuilder()
                .enable(Layer.FHIR, false)
                .enable(Layer.EGYPTIAN, false)
                .build();
        assertTrue(handler.enabledLayers().contains(Layer.BEARER));
        assertTrue(handler.enabledLayers().contains(Layer.HEADERS));
        assertEquals(2, handler.enabledLayers().size());
    }

    @Test
    void envelopeMissingPayloadFieldRaisesBusinessException() {
        RecipientHandler handler = validBuilder().build();
        BusinessException ex = assertThrows(BusinessException.class, () ->
                handler.handle("Bearer x",
                        validHeaders(UUID.randomUUID().toString(), Instant.now()),
                        "{\"other\":\"x\"}"));
        assertEquals("ERR-B-ENV-001", ex.getCode());
    }

    @Test
    void envelopeMalformedJsonRaisesBusinessException() {
        RecipientHandler handler = validBuilder().build();
        BusinessException ex = assertThrows(BusinessException.class, () ->
                handler.handle("Bearer x",
                        validHeaders(UUID.randomUUID().toString(), Instant.now()),
                        "not json"));
        assertEquals("ERR-B-ENV-002", ex.getCode());
    }
}

File: sdk-javascript/tests/unit/recipientHandler.test.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import { type KeyLike, generateKeyPair } from 'jose';
import { beforeEach, describe, expect, it } from 'vitest';

import { randomUUID } from 'node:crypto';
import type { BearerTokenValidator } from '../../src/auth/BearerTokenValidator.js';
import { OutboundEncryptor } from '../../src/client/OutboundEncryptor.js';
import { JWE_ALG } from '../../src/crypto/JweAlgorithms.js';
import {
  AuthenticationError,
  BadTimestampError,
  BadUuidError,
  BusinessError,
  EnvelopeMalformedJsonError,
  EnvelopeMissingPayloadError,
  MissingHeaderError,
  NationalIdInvalidError,
  NotABundleError,
  PatientMissingNationalIdError,
  PatientNonEgyptianError,
  PhoneInvalidError,
  RecipientCodeMismatchError,
  TimestampOutOfRangeError,
} from '../../src/exceptions/HfcxError.js';
import {
  PROTOCOL_HEADER_API_CALL_ID,
  PROTOCOL_HEADER_CORRELATION_ID,
  PROTOCOL_HEADER_RECIPIENT_CODE,
  PROTOCOL_HEADER_SENDER_CODE,
  PROTOCOL_HEADER_TIMESTAMP,
} from '../../src/protocol/ProtocolHeaders.js';
import { Layer } from '../../src/recipient/Layer.js';
import type { LocalKeyProvider } from '../../src/recipient/LocalKeyProvider.js';
import { RecipientHandler } from '../../src/recipient/RecipientHandler.js';
import type { ParticipantCert, RecipientCertResolver } from '../../src/registry/ParticipantCert.js';

const LOCAL = 'payerco@hcx-egypt';
const REMOTE = 'myhospital@hcx-egypt';

let publicKey: KeyLike;
let privateKey: KeyLike;
let encryptor: OutboundEncryptor;
let clock: () => Date;

class StaticKeyProvider implements LocalKeyProvider {
  constructor(private readonly key: KeyLike) {}
  async getPrivateKey(): Promise<KeyLike> {
    return this.key;
  }
}

class StubBearerValidator implements BearerTokenValidator {
  validate(authorizationHeader: string | null | undefined): void {
    if (!authorizationHeader || !authorizationHeader.startsWith('Bearer ')) {
      throw new AuthenticationError('Missing or malformed Authorization header');
    }
  }
}

class StaticResolver implements RecipientCertResolver {
  constructor(private readonly cert: ParticipantCert) {}
  async getRecipientCert(): Promise<ParticipantCert> {
    return this.cert;
  }
}

beforeEach(async () => {
  ({ publicKey, privateKey } = await generateKeyPair(JWE_ALG, {
    modulusLength: 2048,
    extractable: true,
  }));
  encryptor = new OutboundEncryptor(
    new StaticResolver({
      participantCode: LOCAL,
      publicKey,
      notAfter: new Date(Date.now() + 3600_000),
    }),
  );
  clock = () => new Date('2026-05-08T12:00:00Z');
});

function buildHandler(layers?: Set<Layer>): RecipientHandler {
  return new RecipientHandler({
    keyProvider: new StaticKeyProvider(privateKey),
    localParticipantCode: LOCAL,
    bearerTokenValidator: new StubBearerValidator(),
    ...(layers !== undefined ? { enabledLayers: layers } : {}),
    clock,
  });
}

function validBundle(): string {
  return JSON.stringify({
    resourceType: 'Bundle',
    type: 'collection',
    entry: [
      {
        resource: {
          resourceType: 'Patient',
          identifier: [
            {
              system: 'http://hcx-egypt.gov.eg/identifiers/national-id',
              value: '29504150112355',
            },
          ],
          address: [{ country: 'EG' }],
        },
      },
    ],
  });
}

async function envelope(bundle: string): Promise<string> {
  const jwe = await encryptor.encrypt(bundle, LOCAL);
  return JSON.stringify({ payload: jwe });
}

function validHeaders(overrides: Partial<Record<string, string>> = {}): Record<string, string> {
  const base = {
    [PROTOCOL_HEADER_SENDER_CODE]: REMOTE,
    [PROTOCOL_HEADER_RECIPIENT_CODE]: LOCAL,
    [PROTOCOL_HEADER_CORRELATION_ID]: randomUUID(),
    [PROTOCOL_HEADER_TIMESTAMP]: clock().toISOString(),
    [PROTOCOL_HEADER_API_CALL_ID]: randomUUID(),
  };
  return { ...base, ...overrides };
}

describe('RecipientHandler — happy path', () => {
  it('all layers enabled returns RecipientResult', async () => {
    const handler = buildHandler();
    const env = await envelope(validBundle());
    const result = await handler.handle('Bearer test', validHeaders(), env);
    expect(result.decryptedPayload).toContain('Bundle');
    expect(result.correlationId).toBeDefined();
  });

  it('all layers disabled still decrypts', async () => {
    const handler = new RecipientHandler({
      keyProvider: new StaticKeyProvider(privateKey),
      enabledLayers: new Set(),
    });
    const env = await envelope(validBundle());
    const result = await handler.handle(null, validHeaders(), env);
    expect(result.decryptedPayload).toContain('Bundle');
  });

  it('defaults to all four layers', () => {
    const handler = buildHandler();
    expect(handler.enabledLayers.size).toBe(4);
  });
});

describe('RecipientHandler — BEARER layer', () => {
  it('missing header raises AuthenticationError', async () => {
    const handler = buildHandler();
    const env = await envelope(validBundle());
    await expect(handler.handle(null, validHeaders(), env)).rejects.toBeInstanceOf(
      AuthenticationError,
    );
  });

  it('layer enabled but no validator → fails at construction', () => {
    expect(
      () =>
        new RecipientHandler({
          keyProvider: new StaticKeyProvider(privateKey),
          localParticipantCode: LOCAL,
          enabledLayers: new Set([Layer.BEARER, Layer.HEADERS]),
        }),
    ).toThrow(TypeError);
  });

  it('disabled bearer → null Authorization is fine', async () => {
    const handler = new RecipientHandler({
      keyProvider: new StaticKeyProvider(privateKey),
      localParticipantCode: LOCAL,
      enabledLayers: new Set([Layer.HEADERS, Layer.FHIR, Layer.EGYPTIAN]),
      clock,
    });
    const env = await envelope(validBundle());
    const result = await handler.handle(null, validHeaders(), env);
    expect(result).toBeDefined();
  });
});

describe('RecipientHandler — HEADERS layer', () => {
  it('recipient mismatch raises', async () => {
    const handler = buildHandler();
    const env = await envelope(validBundle());
    await expect(
      handler.handle(
        'Bearer test',
        validHeaders({ [PROTOCOL_HEADER_RECIPIENT_CODE]: 'someone-else' }),
        env,
      ),
    ).rejects.toBeInstanceOf(RecipientCodeMismatchError);
  });

  it('bad correlation ID raises', async () => {
    const handler = buildHandler();
    const env = await envelope(validBundle());
    await expect(
      handler.handle(
        'Bearer test',
        validHeaders({ [PROTOCOL_HEADER_CORRELATION_ID]: 'not-a-uuid' }),
        env,
      ),
    ).rejects.toBeInstanceOf(BadUuidError);
  });

  it('timestamp out of range raises', async () => {
    const handler = buildHandler();
    const env = await envelope(validBundle());
    const future = new Date(clock().getTime() + 3600_000).toISOString();
    await expect(
      handler.handle('Bearer test', validHeaders({ [PROTOCOL_HEADER_TIMESTAMP]: future }), env),
    ).rejects.toBeInstanceOf(TimestampOutOfRangeError);
  });

  it('bad timestamp raises', async () => {
    const handler = buildHandler();
    const env = await envelope(validBundle());
    await expect(
      handler.handle(
        'Bearer test',
        validHeaders({ [PROTOCOL_HEADER_TIMESTAMP]: 'not a timestamp' }),
        env,
      ),
    ).rejects.toBeInstanceOf(BadTimestampError);
  });

  it('missing header raises', async () => {
    const handler = buildHandler();
    const env = await envelope(validBundle());
    const headers = validHeaders();
    delete (headers as Record<string, string>)[PROTOCOL_HEADER_TIMESTAMP];
    await expect(handler.handle('Bearer test', headers, env)).rejects.toBeInstanceOf(
      MissingHeaderError,
    );
  });

  it('layer enabled but no localParticipantCode → fails at construction', () => {
    expect(
      () =>
        new RecipientHandler({
          keyProvider: new StaticKeyProvider(privateKey),
          bearerTokenValidator: new StubBearerValidator(),
          enabledLayers: new Set([Layer.BEARER, Layer.HEADERS]),
        }),
    ).toThrow(TypeError);
  });
});

describe('RecipientHandler — FHIR layer', () => {
  it('non-Bundle raises NotABundleError', async () => {
    const handler = buildHandler();
    const env = await envelope(JSON.stringify({ resourceType: 'Patient' }));
    await expect(handler.handle('Bearer test', validHeaders(), env)).rejects.toBeInstanceOf(
      NotABundleError,
    );
  });

  it('Patient without National ID raises', async () => {
    const handler = buildHandler();
    const bundle = JSON.stringify({
      resourceType: 'Bundle',
      type: 'collection',
      entry: [
        {
          resource: { resourceType: 'Patient', address: [{ country: 'EG' }] },
        },
      ],
    });
    const env = await envelope(bundle);
    await expect(handler.handle('Bearer test', validHeaders(), env)).rejects.toBeInstanceOf(
      PatientMissingNationalIdError,
    );
  });

  it('Patient.address[0].country != EG raises', async () => {
    const handler = buildHandler();
    const bundle = JSON.stringify({
      resourceType: 'Bundle',
      type: 'collection',
      entry: [
        {
          resource: {
            resourceType: 'Patient',
            identifier: [
              {
                system: 'http://hcx-egypt.gov.eg/identifiers/national-id',
                value: '29504150112355',
              },
            ],
            address: [{ country: 'US' }],
          },
        },
      ],
    });
    const env = await envelope(bundle);
    await expect(handler.handle('Bearer test', validHeaders(), env)).rejects.toBeInstanceOf(
      PatientNonEgyptianError,
    );
  });
});

describe('RecipientHandler — EGYPTIAN layer', () => {
  it('invalid National ID raises', async () => {
    const handler = buildHandler();
    const bundle = JSON.stringify({
      resourceType: 'Bundle',
      type: 'collection',
      entry: [
        {
          resource: {
            resourceType: 'Patient',
            identifier: [
              {
                system: 'http://hcx-egypt.gov.eg/identifiers/national-id',
                value: '99999999999991',
              },
            ],
            address: [{ country: 'EG' }],
          },
        },
      ],
    });
    const env = await envelope(bundle);
    await expect(handler.handle('Bearer test', validHeaders(), env)).rejects.toBeInstanceOf(
      NationalIdInvalidError,
    );
  });

  it('invalid phone raises', async () => {
    const handler = buildHandler();
    const bundle = JSON.stringify({
      resourceType: 'Bundle',
      type: 'collection',
      entry: [
        {
          resource: {
            resourceType: 'Patient',
            identifier: [
              {
                system: 'http://hcx-egypt.gov.eg/identifiers/national-id',
                value: '29504150112355',
              },
            ],
            address: [{ country: 'EG' }],
            telecom: [{ system: 'phone', value: '01312345678' }],
          },
        },
      ],
    });
    const env = await envelope(bundle);
    await expect(handler.handle('Bearer test', validHeaders(), env)).rejects.toBeInstanceOf(
      PhoneInvalidError,
    );
  });
});

describe('RecipientHandler — envelope errors', () => {
  it('missing payload raises', async () => {
    const handler = buildHandler();
    await expect(handler.handle('Bearer test', validHeaders(), '{}')).rejects.toBeInstanceOf(
      EnvelopeMissingPayloadError,
    );
  });

  it('malformed JSON raises', async () => {
    const handler = buildHandler();
    await expect(handler.handle('Bearer test', validHeaders(), 'not-json')).rejects.toBeInstanceOf(
      EnvelopeMalformedJsonError,
    );
  });
});

describe('RecipientHandler — typed-subclass catch', () => {
  it('BusinessError catches PatientMissingNationalIdError', async () => {
    const handler = buildHandler();
    const bundle = JSON.stringify({
      resourceType: 'Bundle',
      type: 'collection',
      entry: [{ resource: { resourceType: 'Patient', address: [{ country: 'EG' }] } }],
    });
    const env = await envelope(bundle);
    const err = await handler.handle('Bearer test', validHeaders(), env).catch((e) => e as Error);
    expect(err).toBeInstanceOf(BusinessError);
    expect((err as PatientMissingNationalIdError).code).toBe('ERR-B-004');
  });
});
